### PR TITLE
implement LoggedException in vsc.utils.exceptions module

### PR DESCRIPTION
**Pull request by boegel**
implement idea for self-logging exception mentioned by @riccardomurri in https://github.com/hpcugent/easybuild-framework/pull/1183

### CODE REVIEW

File: lib/vsc/utils/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Get logger defined in caller's environment
+    @return: logger instance (or None if none was found)
+    """
+    logger_cls = logging.getLoggerClass()
+    frame = inspect.currentframe()
+    logger = None
+    try:
+        # frame may be None, see https://docs.python.org/2/library/inspect.html#inspect.currentframe
+        framerecords = (frame is None and []) or inspect.getouterframes(frame)
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: 4 lines please

**Comment by boegel** [comment]
removed the Py2.4 tests

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by boegel** [comment]
@stdweird: remark fixed, was that the only one?

File: lib/vsc/utils/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    Get logger defined in caller's environment
+    @return: logger instance (or None if none was found)
+    """
+    logger_cls = logging.getLoggerClass()
+    frame = inspect.currentframe()
+    logger = None
+    try:
+        # frame may be None, see https://docs.python.org/2/library/inspect.html#inspect.currentframe
+        if frame is None:
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: the whole `try/finally` blck is part of the else (or rather the `if is not None`)

File: lib/vsc/utils/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            framerecords = []
+        else:
+            framerecords = inspect.getouterframes(frame)
+
+        for frameinfo in framerecords:
+            bindings = inspect.getargvalues(frameinfo[0]).locals
+            for val in bindings.values():
+                if isinstance(val, logger_cls):
+                    logger = val
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: add `break`?

File: lib/vsc/utils/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        @param logger: logger to use
+        """
+        super(LoggedException, self).__init__(msg)
+
+        # try to use logger defined in caller's environment
+        if logger is None:
+            logger = get_callers_logger()
+            # search can fail, use root logger as a fallback
+            if logger is None:
+                logger = fancylogger.getLogger()
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: make `fancylogger` a class constant. if someone want to override it, it's the easiest way

File: test/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+@author: Kenneth Hoste (Ghent University)
+"""
+import logging
+import os
+import re
+import tempfile
+from unittest import TestLoader, main
+
+from vsc.utils.exceptions import LoggedException
```
Comment (stdweird) [inline test/exceptions.py]: test `get_callers_logger` itself if possible?

**Comment by stdweird** [comment]
some minor remarks added

File: lib/vsc/utils/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        logger.error(msg)
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: also the log function should be configurable with class constant (probably needed in EB)

File: test/exceptions.py (hunk 5 above)
Comment (boegel) [inline test/exceptions.py]: done

**Comment by boegel** [comment]
@stdweird: remarks fixed

File: lib/vsc/utils/exceptions.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    LOGGER_MODULE = fancylogger
+    # name of logging method to use (must accept a single argument of type string, i.e. the log message)
+    LOGGING_METHOD = 'error'
+
+    def __init__(self, msg, logger=None):
+        """
+        Constructor.
+        @param logger: logger to use
+        """
+        super(LoggedException, self).__init__(msg)
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: is there anything in the `__init__` that could cause the logging part below not to be reached? (i guess only a `raise` in the `__init__` can cause that?)
to be safe, i'd move the `super` after the logging block (then there's always logging).
(maybe this is a bit farfetched :smile: )

**Comment by stdweird** [comment]
@boegel one minor remark left.

File: lib/vsc/utils/exceptions.py (hunk 7 above)
Comment (boegel) [inline lib/vsc/utils/exceptions.py]: I doubt there's anything, but it can't hurt to do the logging first, so changed.

**Comment by boegel** [comment]
@stdweird: remark fixed, pls recheck how `LOGGING_METHOD` is defined, I had to change that to redefine it in EasyBuild (to avoid an infinite (re)raise situation with `log.error`)

**Comment by stdweird** [comment]
what do you do now in EB?

**Comment by boegel** [comment]
@stdweird: see https://github.com/hpcugent/easybuild-framework/pull/1218/files#diff-1b40e5d23ba1913560f23697b64533cbR57

File: lib/vsc/utils/exceptions.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    """Exception that logs it's message when it is created."""
+
+    # logger module to use (must provide getLogger() function)
+    LOGGER_MODULE = fancylogger
+    # logging method to use
+    # takes a logger instance and an argument of type string, i.e. the log message
+    # note: 'self' (LoggedException instance) is passed as 1st argument, which is typically just ignored
+    LOGGING_METHOD = lambda _, logger, msg: logger.error(msg)
+
+    def __init__(self, msg, logger=None):
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: support `*args`

File: lib/vsc/utils/exceptions.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+class LoggedException(Exception):
+    """Exception that logs it's message when it is created."""
+
+    # logger module to use (must provide getLogger() function)
+    LOGGER_MODULE = fancylogger
+    # logging method to use
+    # takes a logger instance and an argument of type string, i.e. the log message
+    # note: 'self' (LoggedException instance) is passed as 1st argument, which is typically just ignored
+    LOGGING_METHOD = lambda _, logger, msg: logger.error(msg)
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: why is this needed?

File: lib/vsc/utils/exceptions.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        @param logger: logger to use
+        """
+        # try to use logger defined in caller's environment
+        if logger is None:
+            logger = get_callers_logger()
+            # search can fail, use root logger as a fallback
+            if logger is None:
+                logger = self.LOGGER_MODULE.getLogger()
+
+        self.LOGGING_METHOD(logger, msg)
```
Comment (stdweird) [inline lib/vsc/utils/exceptions.py]:
```python
getartr(logger, self.LOGGING_METHOD, logger.error)(msg, *args)
```

File: lib/vsc/utils/exceptions.py (hunk 9 above)
Comment (boegel) [inline lib/vsc/utils/exceptions.py]: before, this was just `'error'`, but that doesn't work in an EasyBuild context (since the `error` defined by EB raises `EasyBuildError`, which then results in a raise-log infinite loop)

File: lib/vsc/utils/exceptions.py (hunk 10 above)
Comment (boegel) [inline lib/vsc/utils/exceptions.py]: no it's not (anymore)

File: lib/vsc/utils/exceptions.py (hunk 8 above)
Comment (boegel) [inline lib/vsc/utils/exceptions.py]: ok, I'll tackle that

File: lib/vsc/utils/exceptions.py (hunk 9 above)
Comment (boegel) [inline lib/vsc/utils/exceptions.py]: see https://github.com/hpcugent/easybuild-framework/pull/1218/files#diff-1b40e5d23ba1913560f23697b64533cbR57, we need to pass a custom method in EasyBuild to avoid the infinite loop

Comment (stdweird) [inline lib/vsc/utils/exceptions.py]: in EB, make a `_log_no_raise` method as part of `EasyBuildLog`, and set `LOGGING_METHOD = '_log_no_raise'`?

Comment (boegel) [inline lib/vsc/utils/exceptions.py]: Hmm, ok, I'll see if that works out as expected/hoped. I *think* there was something that was blocking me from doing that, but I may be mixing things up.

**Comment by boegel** [comment]
@stdweird: remarks fixed, should be good to go? clean approach for specifying logging method just based on name works out well in EasyBuild

**Comment by stdweird** [comment]
@boegel ok, seems sufficient for EB so merging this in. nice work
